core/store/orm/orm.go

```diff
@@ -291,41 +295,123 @@ func (orm *ORM) convenientTransaction(callback func(*gorm.DB) error) error {
+		)
 		if result.Error != nil {
-			if strings.Contains(result.Error.Error(), "duplicate key value violates unique constraint") {
-				return ErrOptimisticUpdateConflict
-			}
-			return result.Error
+			return errors.Wrap(result.Error, "failed to update job run")
 		}
 		if result.RowsAffected == 0 {
 			return ErrOptimisticUpdateConflict
 		}
-		err := dbtx.Unscoped().Save(&run.Result).Error
-		if err != nil {
-			return err
+
+		uRunResultStrs := []string{}
+		uRunResultArgs := []interface{}{}
+		if run.Result.ID > 0 {
+			uRunResultStrs = append(uRunResultStrs, "(?::bigint, ?::jsonb, ?::text)")
+			uRunResultArgs = append(uRunResultArgs, run.ResultID, run.Result.Data, run.Result.ErrorMessage)
+		} else {
+			if run.ResultID.Valid {
+				return errors.Errorf("got valid ResultID %v for job run %s but Result.ID was 0", run.ResultID.Int64, run.ID)
+			}
+			// Insert the run result for the job run
+			err := dbtx.Exec(`
+WITH run_result AS (
+	INSERT INTO run_results (data, error_message, created_at, updated_at) VALUES(?, ?, NOW(), NOW()) RETURNING id
+)
+UPDATE job_runs
+SET result_id = run_result.id
+FROM run_result
+WHERE job_runs.id = ?
+`, run.Result.Data, run.Result.ErrorMessage, run.ID).Error
+			if err != nil {
+				return errors.Wrap(err, "error inserting run result for job_run")
+			}
 		}
-		if len(run.TaskRuns) > 0 {
-			err = dbtx.Unscoped().Save(&run.TaskRuns).Error
+
+		uTaskRunStrs := []string{}
+		uTaskRunArgs := []interface{}{}
+		for _, tr := range run.TaskRuns {
+			uTaskRunStrs = append(uTaskRunStrs, "(?::uuid, ?::run_status)")
+			uTaskRunArgs = append(uTaskRunArgs, tr.ID, tr.Status)
+			if tr.Result.ID > 0 {
+				uRunResultStrs = append(uRunResultStrs, "(?::bigint, ?::jsonb, ?::text)")
+				uRunResultArgs = append(uRunResultArgs, tr.ResultID, tr.Result.Data, tr.Result.ErrorMessage)
+			} else {
+				if tr.ResultID.Valid {
+					return errors.Errorf("got valid ResultID %v for task run %s but Result.ID was 0", tr.ResultID.Int64, tr.ID)
+				}
+				// Insert the run result for the task run
+				err := dbtx.Exec(`
+WITH run_result AS (
+	INSERT INTO run_results (data, error_message, created_at, updated_at) VALUES(?, ?, NOW(), NOW()) RETURNING id
+)
+UPDATE task_runs
+SET result_id = run_result.id
+FROM run_result
+WHERE task_runs.id = ?
+`, tr.Result.Data, tr.Result.ErrorMessage, tr.ID).Error
+				if err != nil {
+					return errors.Wrap(err, "error inserting run result for job_run")
+				}
+			}
+		}
+
+		if len(uTaskRunStrs) > 0 {
+			updateTaskRunsSQL := `
+UPDATE task_runs SET status=updates.status, updated_at=NOW()
+FROM (VALUES
+%s
+) AS updates(id, status)
+WHERE task_runs.id = updates.id
+`
+			/* #nosec G201 */
+			stmt := fmt.Sprintf(updateTaskRunsSQL, strings.Join(uTaskRunStrs, ","))
+			err := dbtx.Exec(stmt, uTaskRunArgs...).Error
 			if err != nil {
-				return err
+				return errors.Wrap(err, "failed to update task runs")
 			}
 		}
-		return nil
+
+		if len(uRunResultStrs) > 0 {
+			updateTaskRunResultsSQL := `
```

Comment:
this is actually a mix of job run and task run updates right? I think we need a better var name and a comment about this polymorphism